### PR TITLE
Allow coercions in the CountIf function

### PR DESCRIPTION
**Pull request by CarlosFigueiraMSFT**
Currently the CountIf function disallows coercions to its parameters, but there is nothing that should prevent that from happening. This change removes that restriction and adds tests to validate that coercions work as intended.

This also fixes the CountIf function to work with multiple predicates.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs [hunk 1]
```diff
@@ -76,8 +76,15 @@ public override bool CheckTypes(TexlNode[] args, DType[] argTypes, IErrorContain
                 }
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs:76]: Do we need this separate check for OptionSetValue in that case?

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 2]
```diff
@@ -280,21 +280,30 @@ public void TexlFunctionTypeSemanticsCountA()
                 symbol);
         }
 
-        [Fact]
-        public void TexlFunctionTypeSemanticsCountIf()
+        [Theory]
+        [InlineData("CountIf(Table, A < 10)")]
+        [InlineData("CountIf(Table, A < 10, A > 0, A <> 2)")]
+        [InlineData("CountIf([1,2,3], Value) // Coercion from number to boolean")]
+        [InlineData("CountIf([\"false\",\"true\",\"false\"], Value) // Coercion from text to boolean")]
```
Comment (Shpakh-MSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:287]: >[\"false\",\"true\",\"false\ [](http://example.com/codeflow?start=29&length=28)

Any case for error scenario?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs (hunk 1 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs:76]: We already have a check for OptionSetValue. This extends it to other types.

**Comment by Shpakh-MSFT** [comment]
This change looks to be bigger than just CountIf function as description says.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs (hunk 1 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs:76]: Ah, good question. Let me try it out

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs [hunk 3]
```diff
@@ -70,14 +70,17 @@ public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DTyp
             // Ensure that all the args starting at index 1 are booleans or coerece to boolean if they are OptionSetValues.
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs:70]: >coerece to boolean if they are OptionSetValues [](http://example.com/codeflow?start=76&length=46)

Can you update this comment?

Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/CountIf.cs]: Updated

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 4]
```diff
@@ -292,21 +294,63 @@ public void TexlFunctionTypeSemanticsCountA()
...
+
+            bool IExternalOptionSet.IsBooleanValued => true;
+        }
+
+        [Theory]
+        [InlineData("CountIf(Table, A < 10)")]
+        [InlineData("CountIf(Table, A < 10, A > 0, A <> 2)")]
+        [InlineData("CountIf([1,2,3], Value) // Coercion from number to boolean")]
+        [InlineData("CountIf([\"false\",\"true\",\"false\"], Value) // Coercion from text to boolean")]
+        [InlineData("CountIf([1,2,3], BoolOptionSet.Yes) // Coercion from 2-valued option set to boolean")]
```
Comment (Shpakh-MSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:312]: >BoolOptionSet.Yes [](http://example.com/codeflow?start=38&length=17)

Can you add below case as well?

CountIf([1,2,3], If(true, 0 > 1, "true"))

**Review by Shpakh-MSFT — DISMISSED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**